src/main/java/org/codehaus/mojo/flatten/FlattenMojo.java

```diff
@@ -905,14 +960,111 @@ protected void createFlattenedDependencies( Model effectiveModel, List<Dependenc
+            ArtifactDescriptorResult artifactDescriptorResult = this.artifactDescriptorReader
+                    .readArtifactDescriptor(this.session.getRepositorySession(), request);
+
+            for (org.eclipse.aether.graph.Dependency artifactDependency: artifactDescriptorResult.getDependencies()) {
+                if ("test".equals(artifactDependency.getScope())) {
+                    continue;
+                }
+                Exclusion exclusion = new Exclusion();
+                exclusion.setGroupId(artifactDependency.getArtifact().getGroupId());
+                exclusion.setArtifactId(artifactDependency.getArtifact().getArtifactId());
+                exclusions.add(exclusion);
+            }
+
+            dependency.setExclusions(exclusions);
+
+            // convert dependency to string for the set, since Dependency doesn't implement equals, etc.
+            String dependencyString = dependency.toString();
+
+            if (processedDependencies.contains(dependencyString)) {
+                continue;
+            }
+
+            processedDependencies.add(dependencyString);
+            Dependency flattenedDependency = createFlattenedDependency( dependency );
+            if (flattenedDependency != null) {
+                flattenedDependencies.add(flattenedDependency);
+            }
+        }
+    }
+
+    /**
+     * Collects the resolved {@link Dependency dependencies} from the given <code>effectiveModel</code>.
+     *
+     * @param effectiveModel is the effective POM {@link Model} to process.
+     * @param flattenedDependencies is the {@link List} where to add the collected {@link Dependency dependencies}.
+     */
+    protected void createFlattenedDependencies( Model effectiveModel, List<Dependency> flattenedDependencies )
+            throws MojoExecutionException
+    {
+        getLog().debug( "Resolving dependencies of " + effectiveModel.getId() );
+        // this.project.getDependencies() already contains the inherited dependencies but also those from profiles
+        // List<Dependency> projectDependencies = currentProject.getOriginalModel().getDependencies();
+        List<Dependency> projectDependencies = effectiveModel.getDependencies();
+
+        if (flattenDependencyMode == null | flattenDependencyMode == FlattenDependencyMode.direct) {
+            createFlattenedDependenciesDirect(projectDependencies, flattenedDependencies);
+        } else if (flattenDependencyMode == FlattenDependencyMode.all) {
+            try {
+                createFlattenedDependenciesAll(projectDependencies, flattenedDependencies);
+            } catch (Exception e) {
+                throw new MojoExecutionException("caught exception when flattening dependencies", e);
+            }
+        }
+    }
+
     /**
      * @param projectDependency is the project {@link Dependency}.
      * @return the flattened {@link Dependency} or <code>null</code> if the given {@link Dependency} is NOT relevant for
```

Comment:
Strange. This existing function does not flatten dependency.

Reply:
the existing function flattens in different way - it doesn't actually recurse into transitive dependencies.